flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/StandardScaler.scala

```diff
@@ -0,0 +1,178 @@
+import org.apache.flink.util.Collector
+
+/** Scales observations, so that all features have mean equal to zero
+  * and standard deviation equal to one
+  *
+  * This transformer takes a a Vector of values and maps it into the
+  * scaled Vector that each feature has a user-specified mean and standard deviation.
+  *
+  * This transformer can be prepended to all [[Transformer]] and
+  * [[org.apache.flink.ml.common.Learner]] implementations which expect an input of
+  * [[Vector]].
+  *
+  * @example
+  * {{{
+  *                              val trainingDS: DataSet[Vector] = env.fromCollection(data)
+  *
+  *                              val transformer = StandardScaler().setMean(10.0).setStd(2.0)
+  *
+  *                              transformer.transform(trainingDS)
+  * }}}
+  *
+  * =Parameters=
+  *
+  * - [[StandardScaler.Mean]]: The mean value of transformed data set; by default equal to 0
+  * - [[StandardScaler.Std]]: The standard deviation of the transformed data set; by default
+  * equal to 1
+  */
+class StandardScaler extends Transformer[Vector, Vector] with Serializable {
+
+  def setMean(mu: Double): StandardScaler = {
+    parameters.add(Mean, mu)
+    this
+  }
+
+  def setStd(std: Double): StandardScaler = {
+    parameters.add(Std, std)
+    this
+  }
+
+  override def transform(input: DataSet[Vector], parameters: ParameterMap):
+  DataSet[Vector] = {
+    val resultingParameters = this.parameters ++ parameters
+    val mean = resultingParameters(Mean)
+    val std = resultingParameters(Std)
+
+    val featureMetrics = extractFeatureMetrics(input)
+
+    input.map(new RichMapFunction[Vector, Vector]() {
+
+      var broadcastMeanSet: Vector = null
+      var broadcastStdSet: Vector = null
+
+      override def open(parameters: Configuration): Unit = {
+        val broadcastedMetrics = getRuntimeContext().getBroadcastVariable[(Vector,
+          Vector)]("broadcastedMetrics").get(0)
+        broadcastMeanSet = broadcastedMetrics._1
+        broadcastStdSet = broadcastedMetrics._2
+      }
+
+      override def map(vector: Vector): Vector = {
+        var myVector = vector.asBreeze
+
+        myVector :-= broadcastMeanSet.asBreeze
+        myVector :/= broadcastStdSet.asBreeze
+        myVector = (myVector :* std) :+ mean
+        return myVector.fromBreeze
+      }
+    }).withBroadcastSet(featureMetrics, "broadcastedMetrics")
+  }
+
+  /** Calculates in one pass over the data the features' mean and standard deviation.
+    * For the calculation of the Standard deviation with one pas over the data,
+    * the Youngs & Cramer algorithm was used
+    *
+    * @param dataSet The data set for which we want to calculate mean and variance
+    * @return DataSet of Tuple2<featuresMeanVector,featuresStdVector>
+    */
+  private def extractFeatureMetrics(dataSet: DataSet[Vector]) = {
+
+    val metrics = dataSet.combineGroup(new GroupCombineFunction[Vector, (Double, Vector, Vector)] {
```

Comment:
Why do we have this additional ```combineGroup``` step here? Can't we simply use a map on dataSet to create the initial value and then do the reduce operation in a single ```reduce``` call? That way we would not have to much code duplication because the ```combineGroup``` and the subsequent ```reduce``` call have very similar logic. Moreover, you don't have to define a class implementing the ```ReduceFunction```. You can directly use an anonymous function.